### PR TITLE
#1702: Different verification of Dataset properties from setup API and from SparkJobs on execution

### PR DESCRIPTION
**Pull request by benedeki**
* Essentiality Mandatory received a boolean parameter allowRun
* Dataset validation enhanced to support the Mandatory essentiality based on its boolean parameter
* Created ValidationKind enumeration with its Spring converter from Spring too
* Added query params to dataset validation API to specify how strict the validation should be
* Spring RestExceptionHandler changed to return 400 on wrong Query param conversion from string
* Essentiality changed to look little more as an enumeration of case objects
* String implicits: getOrElse, coalesce
* UTs
* Small code changes

While 21 files are changed, most of it are just tiny cascade changes. The most significant changes are in `DatasetService` and `essentiality/package`. Then there's the new `ValidationKind` enumeration.

Closes: #1702

Release notes:
Dataset properties essentiality level _Mandatory_ enhanced with parameter `allowRun`. (If set to true _SparkJobs_ task will still execute even if the property is not defined. _Menas_ UI configuration validation will fail though.)

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/enumerations/ValidationKind.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.menas.utils.enumerations
+
+object ValidationKind extends Enumeration{
```
[the remark this answers is not in the document]
Comment (dk1844) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/enumerations/ValidationKind.scala:18]: why `case object`? Agreed on the `ValidationType` naming.

File: menas/src/main/scala/za/co/absa/enceladus/menas/controllers/DatasetController.scala [hunk 2]
```diff
@@ -103,23 +102,22 @@ class DatasetController @Autowired()(datasetService: DatasetService)
...
+        ResponseEntity.created(location).body(Option(dataset))
     }
   }
 
   @GetMapping(Array("/{datasetName}/{datasetVersion}/properties/valid"))
   @ResponseStatus(HttpStatus.OK)
-  def getPropertiesValidation(@PathVariable datasetName: String, @PathVariable datasetVersion: Int): CompletableFuture[Validation] = {
+  def getPropertiesValidation(@PathVariable datasetName: String,
+                              @PathVariable datasetVersion: Int,
+                              @RequestParam(value = "forRun", required = false, defaultValue = "false") forRun: Boolean
```
[the remark this answers is not in the document]
Comment (Zejnilovic) [reply menas/src/main/scala/za/co/absa/enceladus/menas/controllers/DatasetController.scala:117]: I would have a separate API completely, called `<root>/canRun`.
I don't like boolean parameters.

File: menas/src/main/scala/za/co/absa/enceladus/menas/controllers/RestExceptionHandler.scala [hunk 3]
```diff
@@ -43,6 +44,11 @@ class RestExceptionHandler {
 
   private val logger = LoggerFactory.getLogger(this.getClass)
 
+  @ExceptionHandler(value = Array(classOf[NoSuchElementException]))
+  def handleNoSuchElementException(exception: Exception): ResponseEntity[Any] = {
+    ResponseEntity.badRequest().build[Any]()
+  }
+
```
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/controllers/RestExceptionHandler.scala:51]: Not used. It was actually intended as you describe, but didn't work. (Turns out Spring convertor catches the exception and makes it a `TypeMismatchException`.) Then forgotten. Will remove.

File: menas/src/main/scala/za/co/absa/enceladus/menas/controllers/DatasetController.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/controllers/DatasetController.scala:117]: 🤔

File: menas/src/test/scala/za/co/absa/enceladus/menas/integration/controllers/DatasetApiIntegrationSuite.scala [hunk 4]
```diff
@@ -323,6 +343,19 @@ class DatasetApiIntegrationSuite extends BaseRestApiTest with BeforeAndAfterAll
...
+    s"GET $urlWithWrongQueryParams" should {
+      "return 400" when {
+        "when the query parameter has an unrecognized value" in {
+          val response = sendGet[String](urlWithWrongQueryParams)
+          assertBadRequest(response)
+          val actual = response.getBody
+          assertResult(actual)(s"Unrecognized value '$wrongValidateProperties' for parameter `validateProperties`")
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/test/scala/za/co/absa/enceladus/menas/integration/controllers/DatasetApiIntegrationSuite.scala:358]: I would keep it. It also better shows where the wrong value will be displayed.

File: menas/src/main/scala/za/co/absa/enceladus/menas/controllers/DatasetController.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/controllers/DatasetController.scala:117]: Endpoint `/{datasetName}/{datasetVersion}/properties/validForRun`?
Makes it obvious it's related to `valid` endpoint IMO.

File: menas/src/test/scala/za/co/absa/enceladus/menas/integration/controllers/DatasetApiIntegrationSuite.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (dk1844) [reply menas/src/test/scala/za/co/absa/enceladus/menas/integration/controllers/DatasetApiIntegrationSuite.scala:358]: 🤷‍♂️

**Review by dk1844 — DISMISSED** [review]
LGTM (just read the code)

**Review by AdrianOlosutean — DISMISSED** [review]
Read the code

**Review by AdrianOlosutean — APPROVED** [review]

**Review by HuvarVer — APPROVED** [review]
- [ ] code review
- [x] build
- [ ] run
- [ ] test

I tested properties options by creating and importing new properties. Tested run standardization with these options and a combination of these options. Tested UI by creating datasets end editing them.
I saw there may be one inconsistency - I can not create a dataset in UI without properties set but I can create a new one by API create or ImportItem.

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala [hunk 5]
```diff
@@ -16,8 +16,8 @@
 package za.co.absa.enceladus.menas.utils.converters
 
 import org.springframework.core.convert.converter.Converter
-import za.co.absa.enceladus.menas.utils.enumerations.ValidationLevel
-import za.co.absa.enceladus.menas.utils.enumerations.ValidationLevel.ValidationLevel
+import za.co.absa.enceladus.utils.validation.ValidationLevel.ValidationLevel
```
Comment (AdrianOlosutean) [inline menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala:19]: This seems a bit weird

File: menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala [hunk 6]
```diff
@@ -26,11 +26,11 @@ import za.co.absa.enceladus.model.menas.scheduler.oozie.OozieScheduleInstance
 
 import scala.language.reflectiveCalls
 import DatasetService.RuleValidationsAndFields
-import za.co.absa.enceladus.menas.utils.enumerations.ValidationLevel
-import za.co.absa.enceladus.menas.utils.enumerations.ValidationLevel.ValidationLevel
+import za.co.absa.enceladus.utils.validation.ValidationLevel.ValidationLevel
```
Comment (AdrianOlosutean) [inline menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala:29]: Are these used?

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala (hunk 5 above)
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala:19]: What exactly? In what way?

File: menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala (hunk 6 above)
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala:29]: Yes, the `ValidationLevel` is used. Easy to check 😉

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala (hunk 5 above)
Comment (AdrianOlosutean) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala:19]: ValidationLevel.ValidationLevel

Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/StringToValidationKindConverter.scala:19]: That's how it looks like when defining enumeration as a descendant of `Enumeration` and not as `case object`s.

**Pull request merged**